packages/services/test/src/utils.ts

```diff
@@ -454,3 +679,119 @@ class TerminalTester extends SocketTester {
   private _onMessage: (msg: TerminalSession.IMessage) => void = null;
 }
 
+
+/**
+ * Test a single emission from a signal.
+ *
+ * @param signal - The signal we are listening to.
+ * @param find - An optional function to determine which emission to test,
+ * defaulting to the first emission.
+ * @param test - An optional function which contains the tests for the emission.
+ * @param value - An optional value that the promise resolves to if it is successful.
+ *
+ * @returns a promise that rejects if the function throws an error (e.g., if an
+ * expect test doesn't pass), and resolves otherwise.
+ *
+ * #### Notes
+ * The first emission for which the find function returns true will be tested in
+ * the test function. If the find function is not given, the first signal
+ * emission will be tested.
+ *
+ * You can test to see if any signal comes which matches a criteria by just
+ * giving a find function. You can test the very first signal by just
+ * giving a test function. And you can test the first signal matching the
+ * find criteria by giving both.
+ */
+export
+async function testEmission<T, U, V>(signal: ISignal<T, U>, options: {
+  find?: (a: T, b: U) => boolean
+  test?: (a: T, b: U) => void,
+  value?: V
+}): Promise<V> {
+  const done = new PromiseDelegate<V>();
+  const object = {};
+  signal.connect((sender: T, args: U) => {
+    if (!options.find || options.find(sender, args)) {
+      try {
+        Signal.disconnectReceiver(object);
+        if (options.test) {
+          options.test(sender, args);
+        }
+      } catch (e) {
+        done.reject(e);
+        // throw e;
+      }
+      done.resolve(options.value || undefined);
+    }
+  }, object);
+  return done.promise;
+}
+
+/**
+ * Test that promises resolve in order.
+ *
+ * @param promises - a list of promises in the order they should resolve
+ * @param resolutions - a list of resolved values expected from the promises
+ *
+ * #### Notes
+ * The promises must resolve to different values (i.e., resolutions must be a
+ * list of unique values) for this function to work.
+ */
+export
+function testResolveOrder<T1, T2, T3, T4, T5, T6, T7>(promises: [PromiseLike<T1>, PromiseLike<T2>, PromiseLike<T3>, PromiseLike<T4>, PromiseLike<T5>, PromiseLike<T6>, PromiseLike<T6>], resolutions: [T1, T2, T3, T4, T5, T6, T7]): Promise<void>;
```

Comment:
A shame that Typescript doesn't have variadic templates.

Reply:
Since we don't actually use this, you recommend taking it out. Oh well.